### DISCUSSION TITLE
UsingSmallerProcessingCanvasWithSwing

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#467] [Show and tell]
The following concept might be useful to some users combining Swing with the default canvas shrunk in size to become just another control.  Normally draw() would be non-functional if the entire canvas is removed from the default window to accommodate Swing components and the user would need a JPanel for drawing.  The technique below obviates the need for a JPanel and allows familiar Processing drawing commands to be used with draw().

```
# Uses Imported mode for py5
from javax.swing import *
from java.awt import *

_wndW = 600
_wndH = 600

def setup():
     size(_wndW,_wndH)
     wnd = get_surface()
     wnd.set_title("Default Processing Window")
     canvas = wnd.get_native()
     frame = canvas.getFrame()
#    Shrink size of canvas so it can be used with draw()
     canvas.setBounds(100,300,120,120)
     ta = JTextArea("Jeremiah was a bullfrog.")
     ta.setBounds(30,30,_wndW - 60,200)
     frame.add(ta)
     ta.repaint()
     global slider
     slider = JSlider(JSlider.HORIZONTAL, 0, 100, 50)
     slider.setBounds(80, 260, 150, 24)
     frame.add(slider)
       
def draw():
     background(0)
     fill(255)
     circle(60,60,slider.getValue())

```
**Output:**
<img width="605" alt="smallCnvs" src="https://github.com/py5coding/py5generator/assets/2294268/af951149-52a1-4088-a6c9-9d24a7b25dac">

**Comment by hx2A** [comment]
@vsquared , this is interesting. What does py5 see as the `width` and `height` of the shrunk canvas? Since you accessed the native window object, I'm not sure if Processing would know that something changed and update those variables as you might want it to do.

**Reply by vsquared** [reply]
Good point, Processing is unaware that the canvas size was changed.  If you print out _width_ and _height_ py5 thinks its the same as the window width, height (ie 600x600 in this example) so I don't think users can use Processing keywords _width_ and _height_ for operations on the reduced size canvas.  If someone wanted to use the actual dimensions of the canvas they would probably have to use a couple of global variables, eg _canvasW, canvasH which would be used to set its bounds and for later operations.

**Reply by hx2A** [reply]
@vsquared , I just had a look at the Processing source code and I have an idea. The `width` and `height` int variables are public in both PApplet and the PGraphics object. I think if you assign the new values to both pairs of variables, the new values will be available for users to use.

**Reply by vsquared** [reply]
Nice work, thanks.  I'll try it.
Oops; Thonny rejected the idea.  Might have to call the variables something like newWidth, newHeight.
<img width="752" alt="wh" src="https://github.com/py5coding/py5generator/assets/2294268/dfe432bd-3484-4487-b86f-af214f832c4c">

**Reply by hx2A** [reply]
Oh, right, that assignment to `width` won't work, even in module mode. It will work if you access the Java objects directly:

```python
get_current_sketch()._instance.width = 120
get_current_sketch()._instance.height = 120
get_current_sketch()._instance.getGraphics().width = 120
get_current_sketch()._instance.getGraphics().height = 120
```
